pravega/zookeeper-operator
# Add hashed configs to pod annotations

### Change log description

- Automatically trigger statefulSet rolling update when config is changed
- Add hashed zookeeper configuration as the pod annotation in the statefulSet

### Purpose of the change

Fixes #454 

### What the code does

Add hashed zookeeper configuration as the pod annotation in the statefulSet.
The makeZkPodAnnotations function computes the hash for the zookeeper config string and calls mergeAnnotations to merge with the annotations from the CR.

There are several possible improvements regarding the patch:
1. I was not sure whether to start the function name with capital case or lower case (e.g. makeZkPodAnnotations or MakeZkPodAnnotations), since both code styles exist in the code base.
2. The `mergeAnnotations` function encodes the same logic as the `mergeLabels` function, we could reduce these two into a general `mergeMap` function to reduce the code redundency.
3. I am not sure if the changes in the `env.sh` could reflect to the zookeeper cluster without a restart. If that also requires a restart, then we can add the hash of result of `makeZkEnvConfigString` to the annotation map too.

### How to verify it

Similar to the steps mentioned in #454 
1. Deploy a default zookeeper cluster
2. Change one of the fields in the config field, and observe that the statefulSet performed a rolling update, and the config used by the zookeeper cluster is updated.
